pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,228 @@
+}
+
+// MaybeTrace will log the message if it is covered by a trace.
+func (r *RaftTracer) MaybeTrace(m raftpb.Message) []*kvserverpb.TracedEntry {
+	switch m.Type {
+	case raftpb.MsgProp, raftpb.MsgApp, raftpb.MsgStorageAppend, raftpb.MsgStorageApply:
+		return r.traceIfCovered(m)
+	case raftpb.MsgAppResp, raftpb.MsgStorageAppendResp, raftpb.MsgStorageApplyResp:
+		r.traceIfPast(m)
+		return nil
+	}
+	return nil
+}
+
+// Close will unregister all the traces and log them. It should be called if the
+// replica is destroyed as it is possible some traces were never closed.
+func (r *RaftTracer) Close() {
+	r.m.Range(func(index kvpb.RaftIndex, t *traceValue) bool {
+		r.numRegistered.Add(-1)
+		r.m.Delete(index)
+		t.logf(5, "cleanup log mark (i%d) during close", index)
+		return true
+	})
+}
+
+// traceIfCovered will log the message if it is covered by a trace. Additionally
+// it returns any saved contexts by index for sending to remote nodes.
+func (r *RaftTracer) traceIfCovered(m raftpb.Message) []*kvserverpb.TracedEntry {
+	var tracedEntries []*kvserverpb.TracedEntry
+	r.m.Range(func(index kvpb.RaftIndex, t *traceValue) bool {
+		switch m.Type {
+		case raftpb.MsgProp, raftpb.MsgApp, raftpb.MsgStorageAppend, raftpb.MsgStorageApply:
+			if inRange(index, m.Entries) {
+				e := t.toTracedEntry(index)
+				tracedEntries = append(tracedEntries, &e)
+				t.logf(4, "%s->%s %v Term:%d Log:%d/%d", raft.DescribeTarget(m.From), raft.DescribeTarget(m.To), m.Type, m.Term, m.LogTerm, m.Index)
+			}
+		}
+		return true
+	})
+	return tracedEntries
+}
+
+// inRange returns true if the index is within the range of the entries. It
+// assumes the entries are sorted by index. Specifically it returns if the index
+// is in the set [first_index,last_index].
+func inRange(index kvpb.RaftIndex, entries []raftpb.Entry) bool {
+	return len(entries) != 0 &&
+		index >= kvpb.RaftIndex(entries[0].Index) && index <= kvpb.RaftIndex(entries[len(entries)-1].Index)
+}
+
+// traceIfPast will log the message if it is a range message the covers a trace.
+func (r *RaftTracer) traceIfPast(m raftpb.Message) {
+	if m.Reject {
+		return
+	}
+	r.m.Range(func(index kvpb.RaftIndex, t *traceValue) bool {
+		shouldLog := false
+		shouldUnregister := false
+		switch m.Type {
+		case raftpb.MsgAppResp, raftpb.MsgStorageAppendResp:
+			shouldLog = kvpb.RaftIndex(m.Index) >= index
+		case raftpb.MsgStorageApplyResp:
+			if len(m.Entries) == 0 {
+				return true
+			}
+			msgIndex := m.Entries[len(m.Entries)-1].Index
+			shouldLog = kvpb.RaftIndex(msgIndex) >= index
+			// We unregister the index here because we are now "done" with
+			// this and don't expect more useful entries. There could still
+			// be non-quorum messages, but they don't impact anything the
+			// client cares about.
+			shouldUnregister = shouldLog
+		}
+		if shouldLog {
+			t.logf(4, "%s->%s %v Term:%d Log:%d/%d", raft.DescribeTarget(m.From), raft.DescribeTarget(m.To), m.Type, m.Term, m.LogTerm, m.Index)
```

Comment:
Same thing re message type-specific logging format.

Reply:
See above - I left this for now, but could change this if you think its useful. I'll add a trace to the bottom of what they look like now in all the places.